nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,125 @@
+  workingDir = "/var/lib/${cfg.stateDir}";
+in
+{
+  options.services.matrix-dendrite = {
+    enable = lib.mkEnableOption "matrix.org dendrite";
+    stateDir = lib.mkOption {
+      type = lib.types.str;
+      default = "matrix-dendrite";
+      description = ''
+        The directory where matrix-dendrite stores its stateful data such as
+        certificates, media and uploads.
+      '';
+    };
+    httpPort = lib.mkOption {
+      type = lib.types.int;
+      default = 8008;
+      description = ''
+        The port to listen for HTTP requests on.
+      '';
+    };
+    httpsPort = lib.mkOption {
+      type = lib.types.int;
+      default = 8448;
+      description = ''
+        The port to listen for HTTPS requests on.
+      '';
+    };
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+        options.global = {
+          server_name = lib.mkOption {
+            type = lib.types.str;
+            example = "example.com";
+            default = config.networking.hostName;
+            description = ''
+              The domain name of the server, with optional explicit port.
+              This is used by remote servers to connect to this server,
+              e.g. matrix.org, localhost:8080, etc.
+              This is also the last part of your UserID.
+            '';
+          };
+          private_key = lib.mkOption {
+            type = lib.types.str;
+            example = "/path/to/matrix_key.pem";
+            default = "matrix_key.pem";
+            description = ''
+              The path to the signing private key file, used to sign
+              requests and events.
+            '';
+          };
+          trusted_third_party_id_servers = lib.mkOption {
+            type = lib.types.listOf lib.types.str;
+            example = [ "matrix.org" ];
+            default = [ "matrix.org" "vector.im" ];
+            description = ''
+              Lists of domains that the server will trust as identity
+              servers to verify third
+            '';
+          };
+        };
+      };
+      default = {};
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite"/>
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.matrix-dendriter-key-generator = {
+      description = "Dendrite Matrix homeserver - Key Generator";
+      wantedBy = [ "matrix-dendrite.service" ];
+      requiredBy = [ "matrix-dendrite.service" ];
+      before = [ "matrix-dendrite.service" ];
+      path = with pkgs; [ matrix-dendrite ];
+      serviceConfig = {
+        Type = "oneshot";
+        DynamicUser = true;
+        StateDirectory = cfg.stateDir;
+        WorkingDirectory = workingDir;
+        RemainAfterExit = true;
+      };
+      script = lib.strings.concatStringsSep " " [
+        "generate-keys"
```

Comment:
I guess this generates a self-signed cert for the homeserver, don't they [have to be signed](https://github.com/matrix-org/synapse/blob/master/docs/MSC1711_certificates_FAQ.md) by a CA to federate?
If so, there's should probably be an option to provide one, or maybe use ACME.

Reply:
Yes, that's a good point. I'll look into using ACME.

Reply:
I see acme requires a reverse proxy set up. Synapse doesn't include this in the service, but just documents how to set everything up. Should I configure an option for setting up acme here or document it?

Reply:
I just grep'd, it looks like all (but web servers ofc) rely on documenting how to use acme. Are we still using docgen, or can I write the documentation in markdown somehow?

Reply:
You're right, with matrix-synapse I'll simply disable tls and let the reverse proxy do it: this is what the [docs](https://github.com/matrix-org/synapse/blob/master/docs/reverse_proxy.md) recommends.
A simple solution could be to add options to specify the paths of the tls keypair and generate the self-signed cert only when tls is enabled but the options are unset. (I don't think there are existing settings in dendrite.yaml because these are passed from the command line)

Instead, the `matrix_key.pem` key is always required. It should only be generated if `global.private_key` or `global.old_private_keys` are unset (see [docs] https://github.com/matrix-org/dendrite/blob/master/docs/serverkeyformat.md).

Reply:
Should we not enable https by default and gate that behind specifying the https port? Or should I make enabling https a separate option?

Reply:
Also, given that old_private_keys can be converted to private_keys, is it worth adding that option? I know the RFC stressed not adding too many.

Reply:
> Should we not enable https by default and gate that behind specifying the https port? Or should I make enabling https a separate option?

Oh right, there's also this: It's not clear to me what these options are doing. Are both http and https enabled by default? How do you disable one listener?

> Also, given that old_private_keys can be converted to private_keys, is it worth adding that option? I know the RFC stressed not adding too many.

Uhm, if you like I think it could be done without defining a new option: 
1. make `private_key` a `types.nullOr types.path` and set it to `null` by default. 
2. set `settings.old_private_keys = lib.mkDefault null` in the `config` section
3. put a gate like this
```
lib.optionalString  (cfg.settings.old_private_keys == null && cfg.settings.private_key == null)
  "--private-key matrix_key.pem"
```

Anyway, this is certainly subjective, but I don't think ~8 options are too many: compare this to the 66 options of matrix-synapse module...

Reply:
Note that Dendrite's readme file also __recommends__ using a reverse proxy like nginx and letting it handle TLS, just like for Synapse: https://github.com/matrix-org/dendrite#requirements

Reply:
@rnhmjoj I addressed the http/https on a local branch that I have yet to push. Now I'm leaning towards remove the https option due to the following:

@pacien I personally will also be using it with a reverse proxy. That being said, maybe we just remove the https option and the user can enable it manually via cfg.settings if they want? There's not a precedent for setting up nginx as part of a module, so I assume it's ok to just document how to do it?

Reply:
So, this is getting a little messy. I just pushed but the derivation currently doesn't work.

These lines: https://github.com/mjlbach/nixpkgs/blob/9b3162789114b2e22fe804932dd37d8e5a01d596/nixos/modules/services/misc/matrix-dendrite.nix#L94-L98

cause errors

```
while evaluating the attribute 'values' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:538:9:
while evaluating the attribute 'values' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:637:7:
while evaluating anonymous function at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:524:28, called from /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:524:17:
while evaluating definitions from `/home/michael/Repositories/nix/nixpkgs/nixos/modules/services/misc/matrix-dendrite.nix':
while evaluating 'dischargeProperties' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:596:25, called from /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:525:137:
while evaluating the attribute 'value' at /home/michael/Repositories/nix/nixpkgs/lib/types.nix:386:60:
value is a string with context while a list was expected, at /home/michael/Repositories/nix/nixpkgs/nixos/modules/services/misc/matrix-dendrite.nix:89:23
```

Also, how do we create namespaces with the new settings RFC?

Reply:
Nvm fixed